File: app/src/main/java/com/ShaHar91/poems/MyApp.java
package com.shahar91.poems;

import android.app.Application;

import com.shahar91.poems.data.InitialRealmData;
import com.shahar91.poems.injection.ApplicationComponent;
import com.shahar91.poems.injection.DaggerApplicationComponent;
import com.shahar91.poems.injection.module.ApplicationModule;
import com.shahar91.poems.networking.NewApiManagerService;

import be.appwise.core.core.CoreApp;
import be.appwise.core.extensions.logging.LoggingExtensionsKt;
import be.appwise.core.networking.NetworkingBuilder;
import io.reactivex.plugins.RxJavaPlugins;
import io.realm.Realm;
import io.realm.RealmConfiguration;

public class MyApp extends Application {
    private static MyApp instance;
    private ApplicationComponent appComponent;

    public static MyApp getInstance() {
        return instance;
    }

    public ApplicationComponent getAppComponent() {
        return this.appComponent;
    }

    @Override
    public void onCreate() {
        super.onCreate();

        instance = this;
        this.appComponent = DaggerApplicationComponent
                .builder()
                .applicationModule(new ApplicationModule(this))
                .build();

        NetworkingBuilder networkBuilder = new NetworkingBuilder()
                .setEndPoint(BuildConfig.API_HOST)
                .setPackageName(getPackageName())
                .setClientSecretValue("")
                .setClientIdValue("")
                .setAppName(getString(R.string.app_name))
                .setVersionName(BuildConfig.VERSION_NAME)
                .setVersionCode(String.valueOf(BuildConfig.VERSION_CODE))
                .setApiVersion("1")
                .setApplicationId(BuildConfig.APPLICATION_ID);

        CoreApp
                .init(this)
                .initializeHawk()
                .initializeLogger(getString(R.string.app_name), BuildConfig.DEBUG)
                .initializeNetworking(networkBuilder, NewApiManagerService.class)
                .initializeErrorActivity(BuildConfig.DEBUG || BuildConfig.FLAVOR == "dev" || BuildConfig.FLAVOR == "stg");

        Realm.init(this);
        RealmConfiguration config = new RealmConfiguration.Builder()
                .deleteRealmIfMigrationNeeded()
                .initialData(new InitialRealmData())
                .build();
        Realm.setDefaultConfiguration(config);

        // Catch Unhandled RxJava Exceptions
        // https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0#error-handling
        RxJavaPlugins.setErrorHandler(throwable -> {
            LoggingExtensionsKt.loge(null, throwable, "");

            if (BuildConfig.DEBUG) {
                // Crash the app in debug mode
                throw new Exception(throwable);
            }
        });
    }
}

File: app/src/main/java/com/ShaHar91/poems/MyApp.kt
package com.shahar91.poems

import android.app.Application
import be.appwise.core.core.CoreApp
import be.appwise.networking.Networking
import com.shahar91.poems.data.database.PoemDatabase
import com.shahar91.poems.data.repositories.AuthRepository
import com.shahar91.poems.data.repositories.CategoryRepository
import com.shahar91.poems.data.repositories.PoemRepository
import com.shahar91.poems.data.repositories.ReviewRepository
import com.shahar91.poems.data.repositories.UserRepository
import com.shahar91.poems.di.KoinInitializer
import com.shahar91.poems.networking.ProtectedRestClient
import com.shahar91.poems.networking.UnProtectedRestClient

class MyApp : Application() {
    companion object {
        lateinit var instance: MyApp
            private set
    }

    override fun onCreate() {
        super.onCreate()
        instance = this

        Networking.Builder(this)
            .setPackageName(packageName)
            .setAppName(getString(R.string.app_name))
            .setApiVersion("1")
            .setVersionCode(BuildConfig.VERSION_CODE.toString())
            .setVersionName(BuildConfig.VERSION_NAME)
            .setClientIdValue("")
            .setClientSecretValue("")
            .build()

        CoreApp.init(this)
            .initializeLogger(getString(R.string.app_name), BuildConfig.DEBUG)
            .initializeErrorActivity(
                BuildConfig.DEBUG || BuildConfig.FLAVOR === "dev" || BuildConfig.FLAVOR === "stg"
            )
            .build()

        KoinInitializer.init(this)
    }
}